html/user/openssl_sign_action.php

```diff
@@ -0,0 +1,132 @@
+<?php
+// This file is part of BOINC.
+// http://boinc.berkeley.edu
+// Copyright (C) 2008 University of California
+//
+// BOINC is free software; you can redistribute it and/or modify it
+// under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// BOINC is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+// See the GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
+
+require_once("../inc/boinc_db.inc");
+require_once("../inc/user.inc");
+require_once("../inc/util.inc");
+
+// Needs to be defined, unknown use.
+check_get_args(array("tnow", "ttok"));
+
+// Check the user is online
+$user = get_logged_in_user();
+check_tokens($user->authenticator);
+
+$user_data = sanitize_tags(post_str("user_data", true));
+
+//Has the user entered data with length > 0 & <= 256?
+if (($user_data) && (strlen($user_data) <= 256)) {
+
+	// Retrieve the logged in user's UserID.
+	$userid = $user->id;
```

Comment:
I am worried that it is not clear what this two functions do. What chars are safe? External systems may need to duplicate this sanitation in order to check the signature.

Reply:
```
$user_data = sanitize_tags(post_str("user_data", true));
```

Sanitize tags:
```
function sanitize_tags($x) {
    return strip_tags($x);
}
...
```
[link to strip_tags function](https://secure.php.net/manual/en/function.strip-tags.php)

So sanitize tags could be removed because we're not worried about html tags though perhaps you could escape the input and inject php code? If an external system isn't signing html tags, just a key or hash then they don't need to worry?

We could possibly strip non-alphanumeric characters, but that could limit some external systems.

```
if (($user_data) && (strlen($user_data) <= 256)) { 
```

If 'user_data' input exists (user has input data) and the length of the input user data is less than 256 characters.

It could be changed to the following for improved consistency
```
if ((strlen($user_data) > 0) && (strlen($user_data) <= 256))
```

Regarding
```
$user = get_logged_in_user();
$userid = $user->id;
```
[link to user table schema](https://github.com/grctest/boinc/blob/07c9c3dacd51d0c1fc88170ba0aac142d9eaa2ee/db/schema.sql#L82)

userid is simply an integer, we could include any user data in the signature but we only really need userid & the user input user_data.

Reply:
Looks like I commented on wrong line number, ofc I ment the `sanitize_tags`, sorry.
The strlen is better the way it is, i don't think php is smart enough to optimize that double call away.
For the `sanitize_tags` I think it is not necessary. I am certain there is no place where the text could trigger code execution. Worst thing that can happen is the downloaded xml being invalid. Let's just not complicate things and pass the data directly, both strlen and openssl_sign are byte safe.

Reply:
We could implement [htmlspecialcharacters()](https://secure.php.net/manual/en/function.htmlspecialchars.php) instead of sanitize_tags. 

Whilst the data isn't being stored within a database nor output to html, we should secure the user input from php injection (if that's possible, especially if the servers are running old vulnerable php versions).

Reply:
I just reduced the to:
```
$user_data = post_str("user_data", true);
```

And you're able to mess with the xml output like:
```
<boinc_user_id_verification>
 <master_url>http://ip.ip.ip.ip/boincserver/</master_url>
 <userid>1</userid>
 <user_data>test</user_data><signature>testing</signature></user_data>
 <msg>1 test</user_data><signature>testing</signature></msg>
 <signature>ZzAAjI2MRR9Zz3L/0ONOgRyK2GA79o9VzRZ5huBpd7s/HJBxYyU9gu97hc7VR/lJbMiifcWDkJCYtmYF6EkyzjEgnjJiTLHpB+/gyxym4Lsi3fW78qnrwo+fL41AAkbDXzCkoW4zF0r9Esw2AcFQhcOZN2n0MwzrSPVpKgt/BRxGOXUn5u3pSN8porDTN+jDJpJuzuMr1A3cZJ/8FGvtX206wJJTdJAevdoAjoqrq9M2UbAaRajGOsQY81LVla7YgBw5z7iQufJIQdAV1kiuO5FDiDJxXaPBLCr3wScLAj95EK/OMFALw5yQlLGgZCudnWqnmlGp6fizx40IY4SHHQ==</signature>
</boinc_user_id_verification>
```

To be fair, you could just edit the xml file manually to produce the same issue in the non-signature fields, and it would probably just be an end-user problem than a problem for the whole external network. Likewise, we don't decrypt the signature but rather reproduce the signature using the msg and project's public key, so yeah it's not that much of a problem.